libraries/state_history/trace_converter.cpp

```diff
@@ -30,25 +38,262 @@ void trace_converter::add_transaction(const transaction_trace_ptr& trace, const
+   bio::filtering_ostreambuf strm;
+   EOS_ASSERT(compression == compression_type::zlib, state_history_exception, "unsupported compression format");
+   strm.push(bio::zlib_decompressor());
+   strm.push(bio::back_inserter(decompressed));
+   bio::write(strm, buffer, len);
+   bio::close(strm);
+   return decompressed;
+}
+
+std::vector<augmented_transaction_trace> prepare_traces(trace_converter&       converter,
+                                                        const block_state_ptr& block_state) {
    std::vector<augmented_transaction_trace> traces;
-   if (onblock_trace)
-      traces.push_back(*onblock_trace);
+   if (converter.onblock_trace)
+      traces.push_back(*converter.onblock_trace);
    for (auto& r : block_state->block->transactions) {
       transaction_id_type id;
       if (r.trx.contains<transaction_id_type>())
          id = r.trx.get<transaction_id_type>();
       else
          id = r.trx.get<packed_transaction>().id();
-      auto it = cached_traces.find(id);
-      EOS_ASSERT(it != cached_traces.end() && it->second.trace->receipt, plugin_exception,
+      auto it = converter.cached_traces.find(id);
+      EOS_ASSERT(it != converter.cached_traces.end() && it->second.trace->receipt, state_history_exception,
                  "missing trace for transaction ${id}", ("id", id));
       traces.push_back(it->second);
    }
-   cached_traces.clear();
-   onblock_trace.reset();
+   converter.cached_traces.clear();
+   converter.onblock_trace.reset();
+   return traces;
+}
+
+template <typename Lambda>
+void for_each_packed_transaction(const eosio::state_history::augmented_transaction_trace& obj, const Lambda& lambda) {
+   auto& trace = *obj.trace;
+   if (trace.failed_dtrx_trace) {
+      for_each_packed_transaction(
+          eosio::state_history::augmented_transaction_trace{trace.failed_dtrx_trace, obj.packed_trx}, lambda);
+   }
+   bool include_packed_trx = obj.packed_trx && !trace.failed_dtrx_trace;
+   if (include_packed_trx) {
+      lambda(*obj.packed_trx);
+   }
+}
+
+template <typename Lambda>
+void for_each_packed_transaction(const std::vector<eosio::state_history::augmented_transaction_trace>& traces,
+                                 const Lambda&                                                         lambda) {
+   for (const auto& v : traces) {
+      for_each_packed_transaction(v, lambda);
+   }
+}
+
+void pack(fc::datastream<char*>& ds, const prunable_data_type&  data, compression_type compression) {
+   if (compression == compression_type::none) {
+      fc::raw::pack(ds, data);
+   }
+   else if (compression == compression_type::zlib) {
+      bytes uncompressed = fc::raw::pack(data);
+      fc::raw::pack(ds, zlib_compress_bytes(uncompressed));
+   }
+}
+
+size_t pack_prunable(bytes& buffer, const packed_transaction& trx,
+                  compression_type compression) {
+
+   const auto& data = trx.get_prunable_data();
+
+   int max_size = data.maximum_pruned_pack_size(compression);
```

Comment:
This not the correct size because for full_legacy it is the size of the signatures & packed_context_free_data but you are going to pack the context_free_segments not the packed_context_free_data.

Reply:
`size_t` instead of `int`

Reply:
code rewritten, no long exists.